README.md
# Podcast Index Client

[![Test](https://github.com/podcastool/podcastdx-client/workflows/Test/badge.svg?branch=master)](https://github.com/podcastool/podcastdx-client/actions/workflows/test.yaml) 
[![npm version](https://badge.fury.io/js/podcast-index-client.svg)](https://badge.fury.io/js/podcast-index-client)
[![codecov](https://codecov.io/gh/podcastool/podcastdx-client/branch/master/graph/badge.svg?token=BW5JV2GQS1)](https://codecov.io/gh/podcastool/podcastdx-client)

The is a client for interacting with <https://podcastindex.org/>. In order for this to be useful, you need to get a developer `API KEY` and `API SECRET` from <https://api.podcastindex.org>

This includes typescript types as well as JSDoc strings that should make interacting with the API easier.

JSON Schema generated from the types contained in the project can be found at <https://github.com/podcastool/podcastdx-client/tree/master/src/schemas>

## Installation

```sh
npm -i podcast-index-client
```

## Usage

Auto-generated docs are available at <https://podcastool.github.io/podcastdx-client/index.html>

```ts
import PodcastIndexClient from "podcast-index-client";

// assumes you have an your key and secret set as environment variables
const client = new PodcastIndexClient({
  key: process.env.API_KEY,
  secret: process.env.API_SECRET,
});

client.search("javascript").then(console.log);

client.recentFeeds().then(console.log);
client.recentNewFeeds().then(console.log);
client.recentEpisodes().then(console.log);

client.podcastByUrl("https://feeds.theincomparable.com/batmanuniversity").then(console.log);
client.podcastById(75075).then(console.log);
client.podcastByItunesId(1441923632).then(console.log);

client.episodesByFeedUrl("https://feeds.theincomparable.com/batmanuniversity").then(console.log);
client.episodesByFeedId(75075).then(console.log);
client.episodesByItunesId(1441923632).then(console.log);
client.episodeById(16795106).then(console.log);
```

## Releasing a new version

1. Ensure you're fully committed
   1. Make sure tests are green schema is correct `yarn tsc && yarn test && yarn validate`
2. Run `yarn publish`

## Community

Join on Discord at <https://discord.gg/d6apPvR3N6> or on <https://podcastindex.social/>


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v6.0.1](https://github.com/podcastool/podcastdx-client/compare/v6.0.0...v6.0.1)

- add coverage [`#2`](https://github.com/podcastool/podcastdx-client/pull/2)
- switch to pnpm [`#1`](https://github.com/podcastool/podcastdx-client/pull/1)
- documentation update for 6.0.0 [`4d252ae`](https://github.com/podcastool/podcastdx-client/commit/4d252ae40cca7bf7b6f691de0b78d19e56ba89c1)
- fix test [`d5a1448`](https://github.com/podcastool/podcastdx-client/commit/d5a144840f0646dc331be4466dc5cacfeee74bc0)
- remove debug lib [`02ceddd`](https://github.com/podcastool/podcastdx-client/commit/02ceddd86384f4e349d17abd2d3bb51ca61791c5)

### [v6.0.0](https://github.com/podcastool/podcastdx-client/compare/v5.0.0...v6.0.0)

> 3 August 2023

- documentation update for 6.0.0 [`2c8e963`](https://github.com/podcastool/podcastdx-client/commit/2c8e96330d92ec921e217bbba9b5ae0ca2bc54c8)
- documentation update for 6.0.0 [`db704c4`](https://github.com/podcastool/podcastdx-client/commit/db704c44fcd00db8039b76e0f4ba5cdf4bfefcf8)
- remove mixpanel [`d2b302c`](https://github.com/podcastool/podcastdx-client/commit/d2b302c84d50e5757dc36c6e0675e809a63ff05e)

### [v5.0.0](https://github.com/podcastool/podcastdx-client/compare/v4.2.0...v5.0.0)

> 10 May 2021

- documentation update for 5.0.0 [`0b028ad`](https://github.com/podcastool/podcastdx-client/commit/0b028ad2d5db4bbd2f3227e5cd0144fdc15bfd02)
- update types and add new search by person endpoint. Also opt-in to tracking by default [`c3c3987`](https://github.com/podcastool/podcastdx-client/commit/c3c3987922cd6684a80d2fc2b837e71a7a91b311)
- version bump 5.0.0 [`9b25a78`](https://github.com/podcastool/podcastdx-client/commit/9b25a78cbf3e28b911b0d357c786839be1f26deb)

#### [v4.2.0](https://github.com/podcastool/podcastdx-client/compare/v4.1.1...v4.2.0)

> 11 March 2021

- documentation update for 4.2.0 [`2ea81b8`](https://github.com/podcastool/podcastdx-client/commit/2ea81b88aa5e600720d32fc3acca3a5e4cc969c0)
- feat: :sparkles: #raw to make direct calls to the index without using a convience method name [`715d52b`](https://github.com/podcastool/podcastdx-client/commit/715d52b2c7bba17409bd60f4a4e986304b5f98c2)
- version bump 4.2.0 [`a9647c2`](https://github.com/podcastool/podcastdx-client/commit/a9647c200321312c39ef6fe37f7e8cd3cd67eb2b)

#### [v4.1.1](https://github.com/podcastool/podcastdx-client/compare/v4.1.0...v4.1.1)

> 1 March 2021

- documentation update for 4.1.1 [`46015d7`](https://github.com/podcastool/podcastdx-client/commit/46015d7d040c836fd890050367bf15e77af1e23d)
- version bump 4.1.1 [`6a05c35`](https://github.com/podcastool/podcastdx-client/commit/6a05c358654980edd0098d2541e02ed5d323a86b)
- chore: :recycle: refactor dev example file [`95e5aaf`](https://github.com/podcastool/podcastdx-client/commit/95e5aaf7e7606fb1e066493657772ae72d5f2285)

#### [v4.1.0](https://github.com/podcastool/podcastdx-client/compare/v4.0.0...v4.1.0)

> 1 March 2021

- documentation update for 4.1.0 [`f76e3af`](https://github.com/podcastool/podcastdx-client/commit/f76e3af35edc022a9efe060b82570eb9251640b1)
- update types based on api updates [`6ae8c6f`](https://github.com/podcastool/podcastdx-client/commit/6ae8c6f9b5fe055cd817b2ef381a24a36db41ff1)
- documentation update for 4.1.0 [`7a2b0a9`](https://github.com/podcastool/podcastdx-client/commit/7a2b0a9fd81c99009eeba069464bab9b49201b2a)

### [v4.0.0](https://github.com/podcastool/podcastdx-client/compare/v3.1.0...v4.0.0)

> 19 January 2021

- documentation update for 3.1.0 [`a7c6e26`](https://github.com/podcastool/podcastdx-client/commit/a7c6e26cf5742f39c98dff4c6e857aab1ff89849)
- documentation update for 4.0.0 [`aa2d602`](https://github.com/podcastool/podcastdx-client/commit/aa2d602097b3595f4c6cd602e68a4c3234a89b1b)
- update script entry [`cf4443c`](https://github.com/podcastool/podcastdx-client/commit/cf4443c9d561af93f13f932bbc4b8dd34055bb57)

#### [v3.1.0](https://github.com/podcastool/podcastdx-client/compare/v3.0.0...v3.1.0)

> 15 January 2021

- documentation update for 3.1.0 [`92f37fe`](https://github.com/podcastool/podcastdx-client/commit/92f37fe4ac5c417619c59490b4389eacc8829902)
- add new episode count property and stats endpoint [`1d1f337`](https://github.com/podcastool/podcastdx-client/commit/1d1f3371fea58e1ab41eb7573f361032a1de5dea)
- version bump 3.1.0 [`b73167e`](https://github.com/podcastool/podcastdx-client/commit/b73167e9260c97273b120ced6f466d806bd6cf75)

### [v3.0.0](https://github.com/podcastool/podcastdx-client/compare/v2.4.1...v3.0.0)

> 18 December 2020

- documentation update for 3.0.0 [`6470b6b`](https://github.com/podcastool/podcastdx-client/commit/6470b6b0b77ae79249160700fd623f663cf4213b)
- create json schema and new tests [`7899104`](https://github.com/podcastool/podcastdx-client/commit/7899104702de4747b7fed69447de9f1297878558)
- update all types [`756182c`](https://github.com/podcastool/podcastdx-client/commit/756182cbf9c2c313e2eede18692d094b98c166de)

#### [v2.4.1](https://github.com/podcastool/podcastdx-client/compare/v2.4.0...v2.4.1)

> 6 October 2020

- documentation update for 2.4.1 [`b26c1b6`](https://github.com/podcastool/podcastdx-client/commit/b26c1b6a2cc7d753f74e31ffd7c1571142b13a7b)
- version bump 2.4.1 [`eb84aff`](https://github.com/podcastool/podcastdx-client/commit/eb84affc39750eacb263e1843c299f8531ae1bb2)
- revert dynamic version loading due to ts build issues [`3ae5a6d`](https://github.com/podcastool/podcastdx-client/commit/3ae5a6da7ff8e4f1c97fa7cdd096fab69b63c51c)

#### [v2.4.0](https://github.com/podcastool/podcastdx-client/compare/v2.3.0...v2.4.0)

> 5 October 2020

- documentation update for 2.4.0 [`3c0a1c0`](https://github.com/podcastool/podcastdx-client/commit/3c0a1c0d23e2bcedd7755c39ad3bf92e4e7f15e8)

src/__test__/recent.test.ts
/* eslint-disable sonarjs/no-duplicate-string */
/* eslint-disable no-await-in-loop */
import PodcastIndexClient from "../index";

describe("recent", () => {
  let client: PodcastIndexClient;

  beforeAll(async () => {
    client = new PodcastIndexClient();
  });

  describe("episodes", () => {
    it("returns a default of 10 episodes", async () => {
      const recentResults = await client.recentEpisodes();
      expect(recentResults.items).toHaveLength(10);
    });

    it("returns a user provided max count of episodes", async () => {
      const max = 20;
      const recentResults = await client.recentEpisodes({ max });
      expect(recentResults.items).toHaveLength(max);
    });

    it("supports exclusion string", async () => {
      const resultToExclude = await client.recentEpisodes();
      const episodeToExclude = resultToExclude.items.find((i) => i.feedLanguage.startsWith("en"));

      const excludedResultsByFeed = await client.recentEpisodes({
        excludeString: episodeToExclude?.feedTitle ?? "Today",
      });

      expect(excludedResultsByFeed.items.map((f) => f.id)).not.toContain(episodeToExclude?.id ?? 0);
      const excludedResultsByEpisode = await client.recentEpisodes({
        excludeString: episodeToExclude?.title ?? "Today",
      });

      expect(excludedResultsByEpisode.items.map((f) => f.id)).not.toContain(
        episodeToExclude?.id ?? 0
      );
    });

    it.skip("supports walking back through episodes string", async () => {
      const first = 1305556075;
      const second = 13421292;
      const third = 3223682;

      const secondEpisode = await client.recentEpisodes({
        before: first,
        max: 1,
      });

      expect(secondEpisode.items[0]).toHaveProperty("id", second);

      const thirdEpisode = await client.recentEpisodes({
        before: second,
        max: 1,
      });

      expect(thirdEpisode.items[0]).toHaveProperty("id", third);
    });
  });

  describe("feeds", () => {
    it("returns a default of 40 feeds", async () => {
      const recentResults = await client.recentFeeds();
      expect(recentResults.feeds).toHaveLength(40);
    });

    it("returns a user provided max count of feeds", async () => {
      const max = 20;
      const recentResults = await client.recentFeeds({ max });
      expect(recentResults.feeds).toHaveLength(max);
    });

    it("returns feeds based on an optional language", async () => {
      const lang = "en";
      const recentResults = await client.recentFeeds({ lang });

      expect(recentResults.feeds.every((f) => f.language === lang)).toEqual(true);
    });

    it("returns feeds based on multiple optional languages", async () => {
      const lang = ["en", "en-us"];
      const recentResults = await client.recentFeeds({ lang });
      expect(recentResults.feeds.every((f) => lang.includes(f.language))).toEqual(true);
    });

    it("returns returns feeds based on since", async () => {
      const latestResults = await client.recentFeeds({ max: 5 });
      const firstResults = await client.recentFeeds({
        since: latestResults.feeds[1].newestItemPublishTime,
      });
      const futureResults = await client.recentFeeds({
        since: latestResults.feeds[0].newestItemPublishTime,
      });

      expect(latestResults.feeds).toHaveLength(5);
      expect(futureResults.feeds).toHaveLength(0);
      expect(firstResults.feeds.length).toBeLessThanOrEqual(1);
    });

    it("returns returns feeds based on single category", async () => {
      const latestResults = await client.recentFeeds({ max: 5 });

      const [category] = latestResults.feeds.reduce<string[]>((acc, curr) => {
        const cats = Object.values(curr.categories ?? {});
        return acc.concat(cats);
      }, []);

      const byCategory = await client.recentFeeds({
        category,
      });

      expect(
        byCategory.feeds.every((f) => Object.values(f.categories ?? {}).includes(category))
      ).toEqual(true);
    });

    it("returns returns feeds based on multiple categories (categories are OR'd)", async () => {
      const latestResults = await client.recentFeeds({ max: 5 });

      const [firstCat, secondCat] = Array.from(
        new Set(
          latestResults.feeds.reduce<string[]>((acc, curr) => {
            const cats = Object.values(curr.categories ?? {});
            return acc.concat(cats);
          }, [])
        )
      );

      const byCategory = await client.recentFeeds({
        category: [firstCat, secondCat],
      });

      expect(
        byCategory.feeds.every((f) => {
          const categories = Object.values(f.categories ?? {});
          return categories.includes(firstCat) || categories.includes(secondCat);
        })
      ).toEqual(true);
    });

    it("returns returns feeds based on single notCategory", async () => {
      const latestResults = await client.recentFeeds({ max: 5 });

      const [category] = latestResults.feeds.reduce<string[]>((acc, curr) => {
        const cats = Object.values(curr.categories ?? {});
        return acc.concat(cats);
      }, []);

      const byCategory = await client.recentFeeds({
        notCategory: category,
      });

      expect(
        byCategory.feeds.every((f) => !Object.values(f.categories ?? {}).includes(category))
      ).toEqual(true);
    });

    it("returns returns feeds based on multiple notCategories", async () => {
      const latestResults = await client.recentFeeds({ max: 5 });

      const [firstCat, secondCat] = Array.from(
        new Set(
          latestResults.feeds.reduce<string[]>((acc, curr) => {
            const cats = Object.values(curr.categories ?? {});
            return acc.concat(cats);
          }, [])
        )
      );

      const byCategory = await client.recentFeeds({
        notCategory: [firstCat, secondCat],
      });

      expect(
        byCategory.feeds.every((f) => {
          const categories = Object.values(f.categories ?? {});
          return !categories.includes(firstCat) && !categories.includes(secondCat);
        })
      ).toEqual(true);
    });
  });

  describe("new feeds", () => {
    it("returns a default of 10 feeds", async () => {
      const recentResults = await client.recentNewFeeds();
      expect(recentResults.feeds).toHaveLength(10);
    });

    it("returns a user provided max count of feeds", async () => {
      const max = 20;
      const recentResults = await client.recentNewFeeds({ max });
      expect(recentResults.feeds).toHaveLength(max);
    });
  });

  describe("new soundbites", () => {
    it("returns a default of 1000 soundbites", async () => {
      const recentResults = await client.recentSoundbites();
      expect(recentResults.items).toHaveLength(1000);
    });
  });
});


src/__test__/search.test.ts
/* eslint-disable no-await-in-loop */
import PodcastIndexClient from "../index";
import { PIApiEpisodeInfo } from "../types";

describe("search api", () => {
  let client: PodcastIndexClient;
  beforeAll(() => {
    client = new PodcastIndexClient();
  });

  it("supports basic query", async () => {
    const defaultCount = 25;
    const searchResult = await client.search("javascript");
    expect(searchResult.count).toEqual(defaultCount);
    expect(searchResult.feeds).toHaveLength(defaultCount);
  });

  it("supports explicit results", async () => {
    const searchResult = await client.search("sexual wellness", { max: 3 });
    expect(searchResult.count).toEqual(3);

    const allEpisodes: PIApiEpisodeInfo[] = [];
    for (let i = 0; i < searchResult.feeds.length; i += 1) {
      const feed = searchResult.feeds[i];
      const feedEpisodes = await client.episodesByFeedId(feed.id);
      allEpisodes.push(...feedEpisodes.items);
    }

    expect(allEpisodes.some((ep) => ep.explicit)).toEqual(true);
  });

  it("supports clean only results", async () => {
    const searchResult = await client.search("sexual wellness", { clean: true, max: 5 });
    expect(searchResult.count).toBeLessThanOrEqual(5);

    expect(searchResult.feeds.some((ep) => ep.explicit)).toEqual(false);
  });

  it("supports max items returned", async () => {
    const max = 2;
    const searchResult = await client.search("javascript", { max });
    expect(searchResult.count).toEqual(max);
    expect(searchResult.feeds).toHaveLength(max);
  });

  it("supports a person query", async () => {
    const searchResult = await client.searchPerson("Adam Curry");
    expect(searchResult.status).toEqual("true");
  });
});


example.ts
import PodcastIndexClient from "./src";

const client = new PodcastIndexClient({
  key: process.env.API_KEY!,
  secret: process.env.API_SECRET!,
});

export async function log(thennable: Promise<unknown>): Promise<void> {
  // eslint-disable-next-line no-console
  console.log(JSON.stringify(await thennable, null, 2));
}

(async function runner() {
  // const result = await client.search("Peace in their time");
  // console.log(JSON.stringify(result));

  // const feed = await client.episodesByFeedId(1333936);
  // console.log(JSON.stringify(feed, null, 2));

  // const foundFeed = await client.podcastByItunesId(1538359974);
  // console.log(JSON.stringify(foundFeed, null, 2));
  // await log(client.recentEpisodes({ max: 3 }));
  // result.items.forEach((f) => console.log(f));
  // result.feeds.forEach((f) => console.log(f.categories));
  // console.log(await client.categories());

  // const episodes = await client.episodesByFeedUrl("https://feeds.buzzsprout.com/217882.rss");
  // const podcast = await client.episodesByFeedUrl("http://feeds.feedburner.com/FreeSexPodcast");
  // console.log(JSON.stringify(feeds));
  // console.log(podcast.items);

  // console.log(JSON.stringify(await client.stats(), null, 2));
  // await log(client.raw("/episodes/byid?id=1258727571"));
  // await log(client.raw("/podcasts/byfeedid?id=75075"));
  // await log(client.raw("/podcasts/byfeedid", { id: 75075 }));
  // const result = await client.recentEpisodes({ max: 25 });
  // console.log(
  //   `in order - ${result.items.reduce(
  //     (inOrder, currentItem, idx, all) =>
  //       inOrder && (idx === 0 || all[idx - 1].datePublished < currentItem.datePublished),
  //     true
  //   )}`
  // );

  await log(client.searchPerson("Adam Curry"));
})();

// 39633 categories
// 386307 categories


src/index.ts
import crypto from "crypto";
import fetch from "node-fetch";
import dotEnv from "dotenv";
import { pick } from "ramda";

import { ApiResponse } from "./types";
import { normalizeKey, toEpochTimestamp } from "./utils";
import { version } from "../package.json";

dotEnv.config();

const clientUserAgent = `podcastdx client/${version}`;
const apiVersion = "1.0";

function encodeObjectToQueryString(qs?: ApiResponse.AnyQueryOptions) {
  if (!qs) {
    return null;
  }

  return Object.entries(qs)
    .map(([key, val]) => {
      if (!val) {
        return null;
      }

      if (Array.isArray(val)) {
        return `${key}[]=${(val as unknown[]).map((v) => encodeURI(`${v}`)).join(",")}`;
      }

      if (val === true) {
        return key;
      }

      return `${key}=${encodeURI(`${val}`)}`;
    })
    .filter((x) => x)
    .join("&");
}

class PodcastIndexClient {
  private apiUrl = `https://api.podcastindex.org/api/1.0`;

  private userAgent = clientUserAgent;

  private version = apiVersion;

  private key: string;

  private secret: string;

  constructor({
    key = process.env.API_KEY,
    secret = process.env.API_SECRET,
  }: {
    key?: string;
    secret?: string;
  } = {}) {
    if (!key || !secret) {
      throw new Error("Unable to initialize due to missing key or secret");
    }
    this.key = key;
    this.secret = secret;
  }

  private generateHeaders() {
    if (!this.key || !this.secret) {
      throw new Error("Missing key or secret");
    }

    const apiHeaderTime = Math.floor(Date.now() / 1000);
    const sha1Algorithm = "sha1";
    const sha1Hash = crypto.createHash(sha1Algorithm);
    const data4Hash = this.key + this.secret + apiHeaderTime;
    sha1Hash.update(data4Hash);
    const hash4Header = sha1Hash.digest("hex");

    return {
      "Content-Type": "application/json",
      "X-Auth-Date": `${apiHeaderTime}`,
      "X-Auth-Key": this.key,
      Authorization: hash4Header,
      "User-Agent": `${this.userAgent}/${this.version}`,
    };
  }

  private fetch<T>(endpoint: string, qs?: ApiResponse.AnyQueryOptions): Promise<T> {
    const queryString = qs ? encodeObjectToQueryString(qs) : null;
    const options = {
      method: `GET`,
      headers: this.generateHeaders(),
    };
    const url = `${this.apiUrl}${endpoint}${queryString ? `?${queryString}` : ``}`;

    return fetch(url, options).then((res) => {
      if (res.status >= 200 && res.status < 300) {
        return res.json();
      }
      throw new Error(res.statusText);
    });
  }

  /**
   * Make a raw request to podcast index. This is an escape hatch for leveraging the auth handling in the client
   * but managing the calls and responses yourself.
   * Example:
   *      client.raw("/podcasts/byfeedid?id=75075");
   *      client.raw("/podcasts/byfeedid", { id: 75075 });
   *
   * @param endpoint
   * @param qs
   */
  public async raw<T>(endpoint: string, qs?: ApiResponse.AnyQueryOptions): Promise<T> {
    return this.fetch<T>(endpoint, qs);
  }

  // #region Search
  /**
   * List all categories
   *
   * @param query search query
   */
  public async categories(): Promise<ApiResponse.Categories> {
    return this.fetch<ApiResponse.Categories>("/categories/list");
  }
  // #endregion

  // #region Search
  /**
   * This call returns all of the feeds that match the search terms in the title, author, or owner of the feed.
   * This is ordered by the last-released episode, with the latest at the top of the results.
   *
   * @param query search query
   */
  public async search(
    query: string,
    options: {
      clean?: boolean;
      max?: number;
      fulltext?: boolean;
    } = {}
  ): Promise<ApiResponse.Search> {
    return this.fetch<ApiResponse.Search>("/search/byterm", {
      q: query,
      max: options.max ?? 25,
      clean: Boolean(options.clean),
      fulltext: Boolean(options.fulltext),
    });
  }

  /**
   * This call returns all of the episodes where the specified person is mentioned.
   *
   * @param query search query
   */
  public async searchPerson(
    query: string,
    options: {
      fulltext?: boolean;
    } = {}
  ): Promise<ApiResponse.SearchPerson> {
    return this.fetch<ApiResponse.SearchPerson>("/search/byperson", {
      q: query,
      fulltext: Boolean(options.fulltext),
    });
  }
  // #endregion

  // #region Recent
  /**
   * This call returns the most recent [max] number of episodes globally across the whole index, in reverse chronological order. Max of 1000
   */
  public async recentEpisodes(
    options: {
      max?: number;
      /** If you pass this argument, any item containing this string will be discarded from the result set. This may, in certain cases, reduce your set size below your “max” value. */
      excludeString?: string;
      /** If you pass an episode id, you will get recent episodes before that id, allowing you to walk back through the episode history sequentially. */
      before?: number;
      /** If present, return the full text value of any text fields (ex: description). If not provided, field value is truncated to 100 words. */
      fulltext?: boolean;
    } = {}
  ): Promise<ApiResponse.RecentEpisodes> {
    return this.fetch<ApiResponse.RecentEpisodes>("/recent/episodes", {
      ...options,
      max: options.max ?? 10,
    });
  }

  /**
   * This call returns the most recently feeds in reverse chronological order.
   *
   * @param options additional api options
   */
  public async recentFeeds(
    options: {
      /** Max number of items to return, defaults to 40 */
      max?: number;
      /** You can specify a hard-coded unix timestamp, or a negative integer that represents a number of seconds prior to now. Either way you specify, the search will start from that time and only return feeds updated since then. */
      since?: number;
      /** specifying a language code (like “en”) will return only feeds having that specific language. */
      lang?: string | string[];
      /** You can pass multiple of these to form an array. The category ids given will be excluded from the result set. */
      notCategory?: string[] | number[] | string | number;
      /** You can pass multiple of these to form an array. It will take precedent over the notCategory[] array, and instead only show you feeds with those categories in the result set. These values are OR'd */
      category?: string[] | number[] | string | number;
    } = {}
  ): Promise<ApiResponse.RecentFeeds> {
    const apiOptions: Record<string, string | number | undefined> = {
      max: options.max ?? 40,
      ...pick(["since"], options),
    };

    if (options.lang) {
      if (Array.isArray(options.lang)) {
        apiOptions.lang = options.lang.join(",");
      } else {
        apiOptions.lang = options.lang;
      }
    }
    if (options.notCategory) {
      apiOptions.notcat = Array.isArray(options.notCategory)
        ? options.notCategory.join(",")
        : options.notCategory;
    }

    if (options.category) {
      apiOptions.cat = Array.isArray(options.category)
        ? options.category.join(",")
        : options.category;
    }

    const result = await this.fetch<ApiResponse.RecentFeeds>("/recent/feeds", apiOptions);

    return {
      ...result,
      feeds: result.feeds
        .map((feed) => {
          if (!feed.categories) {
            return { ...feed, categories: {} };
          }
          return feed;
        })
        .map((feed) => normalizeKey((lang) => lang.toLowerCase(), "language", feed)),
    };
  }

  /**
   * This call returns every new feed added to the index over the past 24 hours in reverse chronological order. Max of 1000
   *
   * @param options
   */
  public async recentNewFeeds(
    options: {
      /** Max number of items to return, defaults to 10 */
      max?: number;
    } = {}
  ): Promise<ApiResponse.RecentNewFeeds> {
    return this.fetch<ApiResponse.RecentNewFeeds>("/recent/newfeeds", {
      max: options.max ?? 10,
    });
  }

  /**
   * The most recent 60 soundbites that the index has discovered
   */
  public async recentSoundbites(): Promise<ApiResponse.RecentSoundbites> {
    return this.fetch<ApiResponse.RecentSoundbites>("/recent/soundbites");
  }
  // #endregion

  // #region Podcasts
  /** This call returns everything we know about the feed. */
  public async podcastByUrl(url: string): Promise<ApiResponse.PodcastByUrl> {
    const result = await this.fetch<ApiResponse.PodcastByUrl>("/podcasts/byfeedurl", { url });
    if (!result.feed.categories) {
      result.feed.categories = {};
    }

    return result;
  }

  /** This call returns everything we know about the feed. */
  public async podcastById(id: number): Promise<ApiResponse.PodcastById> {
    const result = await this.fetch<ApiResponse.PodcastById>("/podcasts/byfeedid", { id });
    if (!result.feed.categories) {
      result.feed.categories = {};
    }

    return result;
  }

  /** If we have an itunes id on file for a feed, then this call returns everything we know about that feed. */
  public async podcastByItunesId(id: number): Promise<ApiResponse.PodcastByItunesId> {
    const result = await this.fetch<ApiResponse.PodcastByItunesId>("/podcasts/byitunesid", { id });
    if (!result.feed.categories) {
      result.feed.categories = {};
    }

    return result;
  }
  // #endregion

  // #region Episodes
  /** This call returns all the episodes we know about for this feed, in reverse chronological order. */
  public async episodesByFeedUrl(
    url: string,
    options: {
      /** You can specify a maximum number of results to return */
      max?: number;
      /** You can specify a hard-coded unix timestamp, or a negative integer that represents a number of seconds prior to right now. Either way you specify, the search will start from that time and only return feeds updated since then. */
      since?: number;
      fulltext?: boolean;
    } = {}
  ): Promise<ApiResponse.EpisodesByFeedUrl> {
    const { since, ...rest } = options;
    return this.fetch<ApiResponse.EpisodesByFeedUrl>("/episodes/byfeedurl", {
      ...rest,
      since: toEpochTimestamp(since),
      url,
    });
  }

  /**
   * This call returns all the episodes we know about for this feed, in reverse chronological order.
   * Note: The id parameter is the internal Podcastindex id for this feed.
   */
  public async episodesByFeedId(
    id: number | number[],
    options: {
      /** You can specify a maximum number of results to return */
      max?: number;
      /** You can specify a hard-coded unix timestamp, or a negative integer that represents a number of seconds prior to right now. Either way you specify, the search will start from that time and only return feeds updated since then. */
      since?: number;
      fulltext?: boolean;
    } = {}
  ): Promise<ApiResponse.EpisodesByFeedId> {
    const { since, ...rest } = options;
    const parsedId = Array.isArray(id) ? id.join(",") : id;
    return this.fetch<ApiResponse.EpisodesByFeedId>("/episodes/byfeedid", {
      ...rest,
      since: toEpochTimestamp(since),
      id: parsedId,
    });
  }

  /**
   * If we have an itunes id on file for a feed, then this call returns all the episodes we know about for the feed, in reverse chronological order.
   * Note: The itunes id parameter can either be the number alone, or be prepended with “id”.
   */
  public async episodesByItunesId(
    id: number,
    options: {
      /** You can specify a maximum number of results to return */
      max?: number;
      /** You can specify a hard-coded unix timestamp, or a negative integer that represents a number of seconds prior to right now. Either way you specify, the search will start from that time and only return feeds updated since then. */
      since?: number | Date;
      fulltext?: boolean;
    } = {}
  ): Promise<ApiResponse.EpisodesByItunesId> {
    const { since, ...rest } = options;
    return this.fetch<ApiResponse.EpisodesByItunesId>("/episodes/byitunesid", {
      ...rest,
      since: toEpochTimestamp(since),
      id,
    });
  }

  /**
   * This call returns a random batch of [max] episodes, in no specific order.
   *
   * Note: If no [max] is specified, the default is 1. You can return up to 40 episodes at a time.
   * Note: Language and category names are case-insensitive.
   * Note: You can mix and match the cat and notcat filters to fine tune a very specific result set.
   */
  public async episodesRandom(
    options: {
      /** You can specify a maximum number of results to return */
      max?: number;
      /** Specifying a language code (like "en") will return only episodes having that specific language. You can specify multiple languages by separating them with commas. If you also want to return episodes that have no language given, use the token "unknown". (ex. en,es,ja,unknown) */
      lang?: string | string[];
      /** You may use this argument to specify that you ONLY want episodes with these categories in the results. Separate multiple categories with commas. You may specify either the category id or the category name */
      cat?: string | string[];
      /** You may use this argument to specify categories of episodes to NOT show in the results. Separate multiple categories with commas. You may specify either the category id or the category name. */
      notcat?: string | string[];
    } = {}
  ): Promise<ApiResponse.RandomEpisodes> {
    const parsedOptions: Record<string, number | string | undefined> = options.max
      ? { max: options.max }
      : {};

    parsedOptions.lang = Array.isArray(options.lang) ? options.lang.join(",") : options.lang;
    parsedOptions.cat = Array.isArray(options.cat) ? options.cat.join(",") : options.cat;
    parsedOptions.notcat = Array.isArray(options.notcat)
      ? options.notcat.join(",")
      : options.notcat;

    return this.fetch<ApiResponse.RandomEpisodes>("/episodes/random", parsedOptions);
  }

  /** Get all the metadata for a single episode by passing its id. */
  public async episodeById(
    id: number,
    options: { fulltext?: boolean } = {}
  ): Promise<ApiResponse.EpisodeById> {
    return this.fetch<ApiResponse.EpisodeById>("/episodes/byid", { id, ...options });
  }
  // #endregion

  public async stats(): Promise<ApiResponse.Stats> {
    return this.fetch<ApiResponse.Stats>("/stats/current");
  }
}

export = PodcastIndexClient;
